### PR TITLE
[Playlist] No injections when disabled in `brave://settings`.

### PR DESCRIPTION
**Pull request by szilardszaloki**
Resolves https://github.com/brave/brave-browser/issues/39084.

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by sangwoo108 — APPROVED** [review]
LGTM 👍

File: components/playlist/renderer/playlist_render_frame_observer.h [hunk 1]
```diff
@@ -72,6 +77,7 @@ class PlaylistRenderFrameObserver final
               std::vector<v8::Local<v8::Value>> args = {}) const;
   void OnMediaDetected(base::Value::List media);
 
+  IsPlaylistEnabledCallback is_playlist_enabled_callback_;
```
Comment (atuchin-m) [inline components/playlist/renderer/playlist_render_frame_observer.h:80]: nit: const

File: browser/profiles/brave_renderer_updater.cc [hunk 2]
```diff
@@ -86,6 +88,11 @@ BraveRendererUpdater::BraveRendererUpdater(
       base::BindRepeating(&BraveRendererUpdater::UpdateAllRenderers,
                           base::Unretained(this)));
 #endif
+
+  pref_change_registrar_.Add(
+      playlist::kPlaylistEnabledPref,
```
Comment (atuchin-m) [inline browser/profiles/brave_renderer_updater.cc:93]: a possible improvement: could we just add items to a list of prefs and then call `pref_change_registrar_.Add(..)` in a for-loop avoiding copy-pasting?

File: components/playlist/renderer/playlist_render_frame_observer.cc [hunk 3]
```diff
@@ -76,6 +79,10 @@ PlaylistRenderFrameObserver::GetMediaResponder() {
 }
 
 void PlaylistRenderFrameObserver::RunScriptsAtDocumentStart() {
+  if (!is_playlist_enabled_callback_.Run()) {
```
Comment (atuchin-m) [inline components/playlist/renderer/playlist_render_frame_observer.cc:82]: It's not clear why `is_playlist_enabled_callback_` is a callback, not just `bool`.
In fact we can't efficiently apply different `playlist_enabled` value without frame reloading.

Comment (atuchin-m) [reply components/playlist/renderer/playlist_render_frame_observer.cc:82]: Also this approach creates `PlaylistRenderFrameObserver` even the matching setting is disabled.
1) it takes some resources and CPU time on frame creation/destruction
2) in future changes this check could be missed in some new method that would result a similar bug.

IMHO, a reliable way is just to check it **before** creating `PlaylistRenderFrameObserver`.

**Comment by atuchin-m** [comment]
What about add a browser test to verify: feature on, settings off => no inject?

**Review by iefremov — APPROVED** [review]

File: components/playlist/renderer/playlist_render_frame_observer.cc (hunk 3 above)
Comment (szilardszaloki) [reply components/playlist/renderer/playlist_render_frame_observer.cc:82]: Even though the `PlaylistRenderFrameObserver` is not sensitive to changes to the `playlist_enabled` flag, there are cases where the frame observers are not being recreated on a page refresh, which means that for open tabs, changes in the `Enable Playlist` toggle (in `brave://settings`) are not reflected — hence we decided to stick with the current approach.

**Pull request merged**